### ISSUE TITLE
Rewrite trainign loop and forward of models

### ISSUE THREAD

**Issue by francesco-vaselli** [francesco-vaselli/wipfs#2]
Currently the forward method of models such as double_flow implements the opt.step() and loss.backward() methods as well
I would like to move those steps to the training loop along with the writer
This has also the advantage of being able to call torch.no_grad() in the evaluation phase

**Comment by francesco-vaselli** [comment]
I've settled on a compromise by setting the torch.no_grad() inside the forward method if it is called with val==True
